adapters/verizonmedia/verizonmedia.go

```diff
@@ -0,0 +1,189 @@
+
+func (a *VerizonMediaAdapter) SkipNoCookies() bool {
+	return false
+}
+
+func (a *VerizonMediaAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+
+	errs := make([]error, 0, len(request.Imp))
+	if len(request.Imp) == 0 {
+		err := &errortypes.BadInput{
+			Message: "No impression in the bid request",
+		}
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	errors := make([]error, 0, 1)
+
+	var bidderExt adapters.ExtImpBidder
+	err := json.Unmarshal(request.Imp[0].Ext, &bidderExt)
+	if err != nil {
+		err = &errortypes.BadInput{
+			Message: "ext.bidder not provided",
+		}
+		errors = append(errors, err)
+		return nil, errors
+	}
+	var verizonMediaExt openrtb_ext.ExtImpVerizonMedia
+	err = json.Unmarshal(bidderExt.Bidder, &verizonMediaExt)
+	if err != nil {
+		err = &errortypes.BadInput{
+			Message: err.Error(),
+		}
+		errors = append(errors, err)
+		return nil, errors
+	}
+
+	if verizonMediaExt.Dcn == "" {
+		err = &errortypes.BadInput{
+			Message: "Missing param dcn",
+		}
+		errors = append(errors, err)
+		return nil, errors
+	}
+
+	if verizonMediaExt.Pos == "" {
+		err = &errortypes.BadInput{
+			Message: "Missing param pos",
+		}
+		errors = append(errors, err)
+		return nil, errors
+	}
+
+	siteCopy := *request.Site
+	request.Site = &siteCopy
+	changeRequestForBidService(request, &verizonMediaExt)
+	reqJSON, err := json.Marshal(request)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	thisURI := a.URI
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Set("User-Agent", request.Device.UA)
+	headers.Add("x-openrtb-version", "2.5")
+
+	return []*adapters.RequestData{{
+		Method:  "POST",
+		Uri:     thisURI,
+		Body:    reqJSON,
+		Headers: headers,
+	}}, errors
+}
+
+func (a *VerizonMediaAdapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. ", response.StatusCode),
+		}}
+	}
+
+	var bidResp openrtb.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Bad server response: %d. ", err),
+		}}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(1)
+
+	for _, sb := range bidResp.SeatBid {
+		for _, bid := range sb.Bid {
+			if !impIDExists(bid.ImpID, internalRequest.Imp) {
+				return nil, []error{&errortypes.BadServerResponse{
+					Message: fmt.Sprintf("Unknown ad unit code '%s'", bid.ImpID),
+				}}
+			}
+
+			if openrtb_ext.BidTypeBanner != getMediaTypeForImp(bid.ImpID, internalRequest.Imp) {
```

Comment:
Both `getMediaTypeForImp` and `impIDExists` are traversing all the `Imp` array adding overhead over repetitivenes. Can we consolidate them in one function so we only traverses the `Imp` array once?
```
 99 func (a *VerizonMediaAdapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
100 +-- 20 lines: if response.StatusCode == http.StatusNoContent {-----------
120     for _, sb := range bidResp.SeatBid {
121         for _, bid := range sb.Bid {
122             if !impIDExists(bid.ImpID, internalRequest.Imp) {
123                 return nil, []error{&errortypes.BadServerResponse{
124                     Message: fmt.Sprintf("Unknown ad unit code '%s'", bid.ImpID),
125                 }}
126             }
127
128             if openrtb_ext.BidTypeBanner != getMediaTypeForImp(bid.ImpID, internalRequest.Imp) {
129                 //only banner is supported, anything else is ignored
130                 continue
131             }
132
 .
 .
 .
143 func getMediaTypeForImp(impId string, imps []openrtb.Imp) openrtb_ext.BidType {
144     var mediaType openrtb_ext.BidType
145     for _, imp := range imps {
146         if imp.ID == impId {
147             if imp.Banner != nil {
148                 mediaType = openrtb_ext.BidTypeBanner
149             }
150             return mediaType
151         }
152     }
153     return mediaType
154 }
 .
 .
 .
165 func impIDExists(impId string, imps []openrtb.Imp) bool {
166     for _, imp := range imps {
167         if imp.ID == impId {
168             return true
169         }
170     }
171     return false
172 }
```

Reply:
I combined them in 1 function. I'll be available in the next PR.